File: examples/interface/EN.py
#
#  Copyright (c) 2009-2016, Jack Poulson
#  All rights reserved.
#
#  This file is part of Elemental and is under the BSD 2-Clause License, 
#  which can be found in the LICENSE file in the root directory, or at 
#  http://opensource.org/licenses/BSD-2-Clause
#
import El

n0 = 75
n1 = 75
lambda1 = 3
lambda2 = 4
output = False
display = False
worldRank = El.mpi.WorldRank()
worldSize = El.mpi.WorldSize()

# Place a 2D finite-difference matrices next to an identity
# and then make the last column dense
#
# NOTE: Increasing the magnitudes of the off-diagonal entries causes the matrix
#       to become significantly more ill-conditioned and stresses the 
#       effective condition number of the SQD solves.
def ConcatFD2D(N0,N1):
  A = El.DistSparseMatrix()
  height = N0*N1
  width = 2*N0*N1
  A.Resize(height,width)
  localHeight = A.LocalHeight()
  A.Reserve(7*localHeight)
  for sLoc in xrange(localHeight):
    s = A.GlobalRow(sLoc)
    x0 = s % N0
    x1 = s / N0

    # The finite-difference stencil
    A.QueueLocalUpdate( sLoc, s, 15 )
    if x0 > 0:
      A.QueueLocalUpdate( sLoc, s-1,  -1 )
    if x0+1 < N0:
      A.QueueLocalUpdate( sLoc, s+1,   2 )
    if x1 > 0:
      A.QueueLocalUpdate( sLoc, s-N0, -3 )
    if x1+1 < N1:
      A.QueueLocalUpdate( sLoc, s+N0,  4 )

    # The identity
    sRel = s + N0*N1
    A.QueueLocalUpdate( sLoc, sRel,  1  )

    # The dense last column
    A.QueueLocalUpdate( sLoc, width-1, -10/height );

  A.ProcessQueues()
  return A

A = ConcatFD2D(n0,n1)
b = El.DistMultiVec()
El.Gaussian( b, n0*n1, 1 )
if output:
  El.Print( A, "A" )
  El.Print( b, "b" )
if display:
  El.Display( A, "A" )
  El.Display( b, "b" )

if worldRank == 0:
  print('lambda1 = {}, lambda2 = {}'.format(lambda1,lambda2))

ctrl = El.QPAffineCtrl_d()
ctrl.mehrotraCtrl.progress = True
ctrl.mehrotraCtrl.time = True
ctrl.mehrotraCtrl.solveCtrl.progress = True
ctrl.mehrotraCtrl.solveCtrl.time = True

# Solve *with* resolving the regularization
ctrl.mehrotraCtrl.resolveReg = True
startEN = El.mpi.Time()
x = El.EN( A, b, lambda1, lambda2, ctrl )
endEN = El.mpi.Time()
if worldRank == 0:
  print('EN time (resolve reg.): {} seconds'.format(endEN-startEN))
if display:
  El.Display( x, "x" )

# Solve *without* resolving the regularization
ctrl.mehrotraCtrl.resolveReg = False
startEN = El.mpi.Time()
x = El.EN( A, b, lambda1, lambda2, ctrl )
endEN = El.mpi.Time()
if worldRank == 0:
  print('EN time (no resolve reg.): {} seconds'.format(endEN-startEN))
if display:
  El.Display( x, "x" )

xOneNorm = El.EntrywiseNorm( x, 1 )
xTwoNorm = El.Nrm2( x )
e = El.DistMultiVec()
El.Copy( b, e )
El.Multiply( El.NORMAL, -1., A, x, 1., e )
if display:
  El.Display( e, "e" )
eTwoNorm = El.Nrm2( e )
if worldRank == 0:
  print('|| x ||_1       = {}'.format(xOneNorm))
  print('|| x ||_2       = {}'.format(xTwoNorm))
  print('|| A x - b ||_2 = {}'.format(eTwoNorm))

El.Finalize()

File: examples/optimization/EN.cpp
/*
   Copyright (c) 2009-2016, Jack Poulson
   All rights reserved.

   This file is part of Elemental and is under the BSD 2-Clause License,
   which can be found in the LICENSE file in the root directory, or at
   http://opensource.org/licenses/BSD-2-Clause
*/
#include <El.hpp>

El::DistSparseMatrix<double> ConcatFD2D( El::Int n0, El::Int n1 )
{
    El::DistSparseMatrix<double> A;
    const El::Int height = n0*n1;
    const El::Int width = 2*n0*n1;
    A.Resize( height, width );
    const El::Int localHeight = A.LocalHeight();
    A.Reserve( 7*localHeight );

    for( El::Int iLoc=0; iLoc<localHeight; ++iLoc )
    {
        const El::Int i = A.GlobalRow(iLoc);
        const El::Int x0 = i % n0;
        const El::Int x1 = i / n0;

        A.QueueLocalUpdate( iLoc, i, 15 );
        if( x0 > 0 )    A.QueueLocalUpdate( iLoc, i-1,  -1 );
        if( x0+1 < n0 ) A.QueueLocalUpdate( iLoc, i+1,   2 );
        if( x1 > 0 )    A.QueueLocalUpdate( iLoc, i-n0, -3 );
        if( x1+1 < n1 ) A.QueueLocalUpdate( iLoc, i+n0,  4 );

        const El::Int iRel = i + n0*n1;
        A.QueueLocalUpdate( iLoc, iRel, 1 );

        // For now, this is meant to use integer division (for reproducing)
        A.QueueLocalUpdate( iLoc, width-1, -10/height );
    }
    A.ProcessQueues();
    return A;
}

int main( int argc, char* argv[] )
{
    El::Environment env( argc, argv );
    El::mpi::Comm comm = El::mpi::COMM_WORLD;
    const int commRank = El::mpi::Rank( comm );

    try
    {
        const El::Int n0 = El::Input("--n0","height of 2D discretization",50);
        const El::Int n1 = El::Input("--n1","width of 2D discretization",50);
        const double lambda1 = El::Input("--lambda1","one-norm coefficient",3.);
        const double lambda2 = El::Input("--lambda2","two-norm coefficient",4.);
        const bool print = El::Input("--print","print matrices?",false);
        const bool prog = El::Input("--prog","print progress info?",true);
        const bool time = El::Input("--time","time each step of IPM?",true);
        const bool solveProg = El::Input("--solveProg","solver progress?",true);
        const bool solveTime = El::Input("--solveTime","solver timers?",true);
        El::ProcessInput();
        El::PrintInputReport();

        auto A = ConcatFD2D( n0, n1 );
        El::DistMultiVec<double> b;
        El::Gaussian( b, n0*n1, 1 );
        if( print )
        {
            El::Print( A, "A" );
            El::Print( b, "b" );
        }

        El::qp::affine::Ctrl<double> ctrl;
        ctrl.mehrotraCtrl.print = prog;
        ctrl.mehrotraCtrl.time = time;
        ctrl.mehrotraCtrl.solveCtrl.progress = solveProg;
        ctrl.mehrotraCtrl.solveCtrl.time = solveTime;

        El::DistMultiVec<double> x;
        El::Timer timer;
        El::mpi::Barrier( comm );
        if( commRank == 0 )
            timer.Start();
        El::EN( A, b, lambda1, lambda2, x, ctrl );
        if( commRank == 0 )
            El::Output("EN time: ",timer.Stop()," secs");
        if( print )
            El::Print( x, "x" );
    }
    catch( const std::exception& e ) { El::ReportException(e); }

    return 0;
}
